pact/v3/ffi.py

```diff
@@ -0,0 +1,18 @@
+"""
```

Comment:
Question around the library packaging/design. The FFI is agnostic to the version of any Python API interface, so you may want to consider separating it from a `v3` or `v4` API. I believe the folder is more about representing a major version bump (rather than pact specification versions), in which case, this is probably fine.

I assume the ffi functions will also be kept internal to Pact Python and not part of the public API, and therefore allow the flexibility of moving about/refactoring as needed.

Reply:
So first your question around the `v3` naming. I noticed in the existing PR that the classes have been renamed to include a `V3` suffix (e.g., `ConsumerV3`, `PactV3`, etc.). I personally think that using the same class name, but under a different naming scope will better suit Python. This gives the end user a fair bit of flexibility, with all of the below being possibility:

```python
### What I would likely use
from pact import Consumer
import pact.v3 as pv3
Consumer(...)
pv3.Consumer(...)

### but one could also rename on import
from pact import Consumer as ConsumerV2
from pact.v3 import Consumer as ConsumerV3
ConsumerV2(...)
ConsumerV3(...)

### Or if someone wants to just use v3
from pact.v3 import Consumer
```

And the version naming here is specific to the Pact Python version (currently at 2, and it would migrate to 3 when this change is fully merged). It is not a reference to the Pact specification. I assume that the Pact reference library supports both v3 and v4 of the Pact specification.

---

As to your 2nd question about around public/internal functions (Python has no concept of private unfortunately). By convention, a prefix `_` indicates that a module/function is private.

The `pact.v3._ffi` module contains the binary extension which is automatically generated based on the upstream library. It is generally quite difficult to use these, and requires a lot of casting between Python and C types.

The `pact.v3.ffi` will contain the Python-wrapped functions from the `_ffi` modules. I doubt many people would use these functions, but the naming technically means they are public. The reasoning here is that I am assuming these would be stable in as far as the upstream Pact library is stable. If an upstream function is experimental or deprecated, I can quite easily create the appropriate warnings.

Reply:
Thanks. That clarifies it for me.

I agree, having versions in the types themselves is distasteful - I regret having that in the JS library.

> I assume that the Pact reference library supports both v3 and v4 of the Pact specification.

That's correct. However as you'll see, there are different capabilities that come with each version of the spec (particularly on the consumer side). As long as your consumer API is able to constrain the options based on which version of the spec they intend to use, you should be OK.

For example, you can't say "I want to use V2 of the spec" and then go on to define a synchronous message interaction (introduced in V4).  

If you'd like to see that expressed as types (and using the type-state pattern), [this](https://github.com/pact-foundation/pact-js/blob/master/src/v4/http/types.ts) might be helpful. Obviously, the [compatibility suite](https://github.com/pact-foundation/pact-compatibility-suite/) would be a good reference also.

Reply:
As I work through to implementing the entire compatibility suite, I'll make sure that the correct Pact spec is available, and can be constrained :)